### ISSUE TITLE
How does geth communicate with prysm?

### ISSUE THREAD

**Issue by lcgogo** [ethereum/go-ethereum#25594]
Now I run geth 1.10.23 and prysm 3.0.0 with prysm connected to geth 8551 port by a jwt.hex file

But in geth 1.10.23 logs always print WARN like below

```shell
WARN [08-24|13:51:42.419] Merge is configured, but no beacon client seen. Please ensure you have one available before the transition arrives! eta=3w1d28m
```

**Comment by Neurone** [comment]
Teku [seems to be waiting](https://github.com/ConsenSys/teku/blob/b403f8e524f405cf3905c942c6002f4c08b940f3/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/TerminalPowBlockMonitor.java#L133) for the [Bellatrix fork](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement/) (epoch 144896 – 11:34:47am UTC on Sept 6, 2022) to monitor EL for TTD. Maybe Prysm does the same?

**Comment by Neurone** [comment]
Checked on [Prysm source code](https://github.com/prysmaticlabs/prysm/blob/d077483577bc8fdc658940adf71d8d11dfa5949b/beacon-chain/execution/check_transition_config.go#L29-L39), it seems to me it also wait for Bellatrix fork before starting to monitor EL.

```go
// Checks the transition configuration between Prysm and the connected execution node to ensure
// there are no differences in terminal block difficulty and block hash.
// If there are any discrepancies, we must log errors to ensure users can resolve
//the problem and be ready for the merge transition.
func (s *Service) checkTransitionConfiguration(
	ctx context.Context, blockNotifications chan *feed.Event,
) {
	// If Bellatrix fork epoch is not set, we do not run this check.
	if params.BeaconConfig().BellatrixForkEpoch == math.MaxUint64 {
		return
	}
...
```

**Comment by Neurone** [comment]
Although, it should have been [set that variable few days ago](https://github.com/prysmaticlabs/prysm/blob/dfa400d4a1943da5dc95f716bfcafc196291284f/config/params/mainnet_config.go#L25):

```go
mainnetBellatrixForkEpoch = 144896 // Sept 6, 2022, 11:34:47am UTC
```

So, maybe that is not the issue.

**Comment by Neurone** [comment]
And yes, [it waits too](https://github.com/prysmaticlabs/prysm/blob/d077483577bc8fdc658940adf71d8d11dfa5949b/beacon-chain/execution/check_transition_config.go#L90-L95):

```go
if currentEpoch >= params.BeaconConfig().BellatrixForkEpoch && !hasTtdReached {
	hasTtdReached, err = s.logTtdStatus(ctx, ttd)
	if err != nil {
		log.WithError(err).Error("Could not log ttd status")
	}
}
```

**Comment by lcgogo** [comment]
> 

Thx. I see the printout every 5 mins. Should I wait prysm full synced?
I run geth and prysm in diff docker containers.
Is there any method to test connect from geth side?

**Comment by garyng2000** [comment]
on your prysm beacon, you should see that it has connected to the CL like this

Aug 25 03:36:06 hqprdubeth start_mainnet_beacon.sh[3777890]: time="2022-08-25 03:36:06" level=info msg="Connected to new endpoint: http://localhost:8550" prefix=powchain

**Comment by lcgogo** [comment]
> on your prysm beacon, you should see that it has connected to the CL like this
> 
> Aug 25 03:36:06 hqprdubeth start_mainnet_beacon.sh[3777890]: time="2022-08-25 03:36:06" level=info msg="Connected to new endpoint: http://localhost:8550" prefix=powchain

Understood. Many thanks.

**Comment by ailiao** [comment]
So just to clarify, getting this message is expected until the Bellatrix fork or not?

`WARN [08-24|13:51:42.419] Merge is configured, but no beacon client seen. Please ensure you have one available before the transition arrives! eta=3w1d28m`

**Comment by lcgogo** [comment]
> So just to clarify, getting this message is expected until the Bellatrix fork or not?
> 
> `WARN [08-24|13:51:42.419] Merge is configured, but no beacon client seen. Please ensure you have one available before the transition arrives! eta=3w1d28m`

After correctly config beacon node, this message disappeared. Now I use below option instead of http+port in prysm beacon node
```shell
  --execution-endpoint=/data/geth/geth.ipc
```

**Comment by benesjan** [comment]
Nimbus client behaves the same way and the message is supposed to disappear after Bellatrix fork as well.

A post from dev on Discord:

> For users running the latest execution clients, even though you've configured the authenticated port (using the JWT secret options) you may see a warnings like the following:
> 
> WARN [08-23|09:21:08.395] Merge is configured, but no beacon client seen. Please ensure you have one available before the transition arrives! (geth)
> WARN  | EngineExchangeTransitionConfiguration | not called in 120 seconds, consensus client may not be connected (besu)
> 
> This is normal - Nimbus (v22.8.0) will activate the merge mode as it enters the Bellatrix hard fork on September 6, and the warning will go away. Until then, keep on eye on nimbus logs instead - if there is anything wrong with the connection to geth, a WRN log will appear on the Nimbus side.

**Comment by cortexyphan** [comment]
Bellatrix is dove but this damned written still appears... any suggestion about that?

**Comment by MindPhaser34** [comment]
@cortexyphan had similar behaviour. After bump nimbus version to latest (22.9.0) and restart both (geth+nimbus), message disappeared in geth logs. Try this out and wait for sync nimbus node (in my case was since "slot=4655914 epoch=145497 sync="03d09h48m (62.57%)")

**Comment by nuliknol** [comment]
> We are also getting a similar error, we are on geth 1.10.23 and beacon chain is on prysm 3.1.1 

this message is sent in a loop (every 5 seconds), and it will be sent infinitely until the beacon client connects to Geth.  After the beacon client connects to Geth it will continuously call `ExchangeTransitionConfigurationV1` method via Geth's RPC and this is is how Geth will know that the connection is established and the node is "merge is ready" now, so the message will stop appearing in the log. Check the last date of this message, if it is older than 5 seconds (comparing to current date), then you have merge-ready node.

@Sreeram1993

**Comment by Sreeram1993** [comment]
Thanks @nuliknol for the explanation. I can see that both nodes are syncing and running as expected and is connected via jwt token.  

These are the status from beacon node and geth respectively
--
time="2022-09-13 09:00:23" level=info msg="Ready for The Merge" latestDifficulty=58613127217489670579653 network=mainnet prefix=powchain terminalDifficulty=58750000000000000000000
time="2022-09-13 09:00:24" level=info msg="Processing block batch of size 64 starting from  0x68edd498... 2294500/4686300 - estimated time remaining 34h41m38s" blocksPerSecond=19.1 peers=42 prefix=initial-sync
time="2022-09-13 09:00:28" level=info msg="Processing block batch of size 64 starting from  0x113511d0... 2294564/4686300 - estimated time remaining 34h41m34s" blocksPerSecond=19.1 peers=42 prefix=initial-sync
time="2022-09-13 09:00:31" level=info msg="Processing block batch of size 63 starting from  0x173a31f3... 2294628/4686300 - estimated time remaining 34h46m58s" blocksPerSecond=19.1 peers=43 prefix=initial-sync
time="2022-09-13 09:00:33" level=info msg="Ready for The Merge" latestDifficulty=58613127217489670579653 network=mainnet prefix=powchain terminalDifficulty=58750000000000000000000
time="2022-09-13 09:00:35" level=info msg="Processing block batch of size 64 starting from  0x3dca0bcf... 2294692/4686301 - estimated time remaining 34h46m55s" blocksPerSecond=19.1 peers=44 prefix=initial-sync
time="2022-09-13 09:00:39" level=info msg="Processing block batch of size 62 starting from  0xf12e085d... 2294756/4686301 - estimated time remaining 34h57m50s" blocksPerSecond=19.0 peers=42 prefix=initial-sync
time="2022-09-13 09:00:43" level=info msg="Processing block batch of size 63 starting from  0x2c42f4b0... 2294820/4686301 - estimated time remaining 34h57m47s" blocksPerSecond=19.0 peers=42 prefix=initial-sync
time="2022-09-13 09:00:43" level=info msg="Ready for The Merge" latestDifficulty=58613139221654101626832 network=mainnet prefix=powchain terminalDifficulty=58750000000000000000000
time="2022-09-13 09:00:47" level=info msg="Processing block batch of size 64 starting from  0x7f320bc9... 2294884/4686302 - estimated time remaining 34h57m44s" blocksPerSecond=19.0 peers=41 prefix=initial-sync
--

GETH 
--
INFO [09-13|08:55:13.549] Imported new chain segment               blocks=1 txs=131  mgas=7.941  elapsed=381.470ms mgasps=20.816  number=15,526,007 hash=ebb3bf..c4b429 dirty=0.00B
INFO [09-13|08:55:13.563] Imported new chain segment               blocks=1 txs=0    mgas=0.000  elapsed=6.679ms   mgasps=0.000   number=15,526,008 hash=f04ada..16de23 dirty=0.00B
INFO [09-13|08:55:14.460] Downloader queue stats                   receiptTasks=0 blockTasks=0  itemSize=50.01KiB  throttle=5243
WARN [09-13|08:55:40.722] Merge is configured, but previously seen beacon client is offline. Please ensure it is operational before the transition arrives! eta=2d4h21m
INFO [09-13|08:55:42.737] Imported new chain segment               blocks=1 txs=388  mgas=25.714 elapsed=494.990ms mgasps=51.948  number=15,526,009 hash=5aebc6..fe1a3d dirty=0.00B
INFO [09-13|08:55:54.113] Imported new chain segment               blocks=1 txs=152  mgas=12.513 elapsed=875.794ms mgasps=14.287  number=15,526,010 hash=53c8fa..952d28 dirty=0.00B
INFO [09-13|08:56:14.167] Imported new chain segment               blocks=1 txs=282  mgas=21.815 elapsed=252.148ms mgasps=86.515  number=15,526,011 hash=e68bcc..87f180 dirty=0.00B
INFO [09-13|08:56:50.858] Imported new chain segment               blocks=1 txs=29   mgas=2.316  elapsed=24.590ms  mgasps=94.172  number=15,526,012 hash=86cf33..4e253b dirty=0.00B
INFO [09-13|08:57:14.345] Imported new chain segment               blocks=1 txs=421  mgas=29.987 elapsed=1.681s    mgasps=17.836  number=15,526,013 hash=5dcec2..ee8c7f dirty=0.00B
INFO [09-13|08:58:11.121] Imported new chain segment               blocks=1 txs=109  mgas=30.025 elapsed=873.852ms mgasps=34.359  number=15,526,014 hash=f07db9..1b327e dirty=0.00B
INFO [09-13|08:58:17.060] Downloader queue stats                   receiptTasks=0 blockTasks=0  itemSize=60.48KiB  throttle=4335
INFO [09-13|08:58:17.338] Imported new chain segment               blocks=1 txs=338  mgas=28.791 elapsed=4.324s    mgasps=6.658   number=15,526,015 hash=bf7752..41db88 dirty=0.00B
INFO [09-13|08:58:17.839] Imported new chain segment               blocks=1 txs=431  mgas=5.163  elapsed=500.705ms mgasps=10.311  number=15,526,016 hash=0f2089..034277 dirty=0.00B ignored=1
INFO [09-13|08:58:29.987] Imported new chain segment               blocks=1 txs=277  mgas=27.766 elapsed=751.802ms mgasps=36.932  number=15,526,017 hash=4b57b3..329e4b dirty=0.00B
INFO [09-13|08:58:49.641] Imported new chain segment               blocks=1 txs=178  mgas=13.490 elapsed=394.401ms mgasps=34.205  number=15,526,018 hash=baf163..4154d7 dirty=0.00B
INFO [09-13|08:59:06.287] Imported new chain segment               blocks=1 txs=238  mgas=22.505 elapsed=248.340ms mgasps=90.623  number=15,526,019 hash=e0a2a5..c9be4a dirty=0.00B
INFO [09-13|08:59:12.937] Imported new chain segment               blocks=1 txs=94   mgas=6.724  elapsed=56.823ms  mgasps=118.331 number=15,526,020 hash=10eb76..fda350 dirty=0.00B
INFO [09-13|08:59:21.814] Imported new chain segment               blocks=1 txs=72   mgas=6.118  elapsed=57.861ms  mgasps=105.733 number=15,526,021 hash=223ba6..116f35 dirty=0.00B
INFO [09-13|08:59:34.325] Imported new chain segment               blocks=1 txs=63   mgas=5.692  elapsed=96.257ms  mgasps=59.133  number=15,526,022 hash=03b375..569d06 dirty=0.00B
INFO [09-13|08:59:34.649] Imported new chain segment               blocks=1 txs=0    mgas=0.000  elapsed=6.235ms   mgasps=0.000   number=15,526,023 hash=35a76c..375509 dirty=0.00B
INFO [09-13|08:59:51.907] Imported new chain segment               blocks=1 txs=333  mgas=27.468 elapsed=826.072ms mgasps=33.251  number=15,526,024 hash=a2f1d5..fcb39d dirty=0.00B
INFO [09-13|09:00:34.036] Imported new chain segment               blocks=1 txs=466  mgas=29.992 elapsed=1.627s    mgasps=18.428  number=15,526,025 hash=eba441..937a07 dirty=0.00B
INFO [09-13|09:00:34.342] Imported new chain segment               blocks=1 txs=470  mgas=30.011 elapsed=305.675ms mgasps=98.179  number=15,526,025 hash=3504d9..67f339 dirty=0.00B
WARN [09-13|09:00:40.756] Merge is configured, but previously seen beacon client is offline. Please ensure it is operational before the transition arrives! eta=2d3h38m
INFO [09-13|09:00:50.071] Imported new chain segment               blocks=1 txs=65   mgas=4.620  elapsed=67.104ms  mgasps=68.853  number=15,526,026 hash=ae240a..507b99 dirty=0.00B
INFO [09-13|09:01:04.702] Imported new chain segment
--

Every 5 mins I can see the geth node showing the warning. And is this because the beacon chain is not upto date as well ?

**Comment by LyyDandelion** [comment]
> Yeah you're geth client was not registering the prysm. Question i have is if i am trying to sync geth do i need the conensus layer to finish first before geth is able to process from block 0?

same problem!

**Comment by holiman** [comment]
Communication issues between CL and geth should be fixed by now, this issue has seen no activity for the past three months. 
If you are still having problems with CL-EL comms, please file a new ticket with version-info and logs.
